### PR TITLE
filedata v1.3: inspect smtp attachments

### PR DESCRIPTION
**Pull request by glongo**
This permits to use file_data keyword to inspect smtp attachments.
Comments in the last PR are addressed.

- PR glongo: https://buildbot.openinfosecfoundation.org/builders/glongo/builds/27
- PR glongo-pcap: https://buildbot.openinfosecfoundation.org/builders/glongo-pcap/builds/26

Last PR: https://github.com/inliniac/suricata/pull/1418

@jasonish

### CODE REVIEW

File: src/app-layer-smtp.c [hunk 1]
```diff
@@ -263,6 +264,38 @@ static void SMTPConfigure(void) {
...
+
+    if (t == NULL)
+        return;
+
+    TAILQ_FOREACH(p, &t->head, next) {
+        if (strcasecmp("content-limit", p->name) == 0) {
+            if (ParseSizeStringU32(p->val, &content_limit) < 0) {
+                SCLogError(SC_ERR_SIZE_PARSE, "Error parsing content-limit "
+                           "from conf file - %s. Killing engine", p->val);
+                exit(EXIT_FAILURE);
```
Comment (inliniac) [inline src/app-layer-smtp.c]: Can you remove the exits and do proper error handling? Warning+default value is what I have in mind.

File: src/app-layer-smtp.c [hunk 2]
```diff
@@ -55,6 +55,11 @@
 #include "conf.h"
 
 #include "util-mem.h"
+#include "util-misc.h"
+
+#define FILEDATA_CONTENT_LIMIT 1000
```
Comment (inliniac) [inline src/app-layer-smtp.c]: What is the reason for picking these values? (add comments)

File: src/detect-file-data.c [hunk 3]
```diff
@@ -76,12 +76,104 @@ void DetectFiledataRegister(void)
...
-        SCLogError(SC_ERR_INVALID_SIGNATURE, "Can't use file_data with flow:to_server or from_client with http.");
+
+    if (!DetectProtoContainsProto(&s->proto, IPPROTO_TCP) && s->alproto != ALPROTO_UNKNOWN && s->alproto != ALPROTO_HTTP && s->alproto != ALPROTO_SMTP) {
+        SCLogError(SC_ERR_CONFLICTING_RULE_KEYWORDS, "rule contains conflicting keywords.");
         return -1;
     }
 
-    s->list = DETECT_SM_LIST_HSBDMATCH;
+    if (s->alproto == ALPROTO_HTTP || s->alproto == ALPROTO_SMTP) {
+        if ((s->init_flags & SIG_FLAG_INIT_FLOW) && (s->flags & SIG_FLAG_TOSERVER) && !(s->flags & SIG_FLAG_TOCLIENT)) {
```
Comment (inliniac) [inline src/detect-file-data.c]: Line is too long.

File: src/detect-file-data.c [hunk 4]
```diff
@@ -76,12 +76,104 @@ void DetectFiledataRegister(void)
 static int DetectFiledataSetup (DetectEngineCtx *de_ctx, Signature *s, char *str)
 {
     SCEnter();
-    if ((s->init_flags & SIG_FLAG_INIT_FLOW) && (s->flags & SIG_FLAG_TOSERVER) && !(s->flags & SIG_FLAG_TOCLIENT)) {
-        SCLogError(SC_ERR_INVALID_SIGNATURE, "Can't use file_data with flow:to_server or from_client with http.");
+
+    if (!DetectProtoContainsProto(&s->proto, IPPROTO_TCP) && s->alproto != ALPROTO_UNKNOWN && s->alproto != ALPROTO_HTTP && s->alproto != ALPROTO_SMTP) {
```
Comment (inliniac) [inline src/detect-file-data.c]: Line too long

File: src/detect-file-data.c [hunk 5]
```diff
@@ -76,12 +76,104 @@ void DetectFiledataRegister(void)
...
+        SCLogError(SC_ERR_CONFLICTING_RULE_KEYWORDS, "rule contains conflicting keywords.");
         return -1;
     }
 
-    s->list = DETECT_SM_LIST_HSBDMATCH;
+    if (s->alproto == ALPROTO_HTTP || s->alproto == ALPROTO_SMTP) {
```
Comment (inliniac) [inline src/detect-file-data.c]: Do we need the alproto check here? The behaviour is the same in both smtp and http cases, so we might just as well not check it?

File: src/detect.c [hunk 6]
```diff
@@ -2332,9 +2360,16 @@ static int SignatureCreateMask(Signature *s)
...
-    if (s->sm_lists[DETECT_SM_LIST_HSBDMATCH] != NULL) {
-        s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
-        SCLogDebug("sig requires http app state");
+    if (s->sm_lists[DETECT_SM_LIST_FILEDATA] != NULL) {
+        if (s->alproto == ALPROTO_HTTP)
+            s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
+        else if (s->alproto == ALPROTO_SMTP)
+            s->mask |= SIG_MASK_REQUIRE_SMTP_STATE;
+        else if (s->alproto == ALPROTO_UNKNOWN) {
+            s->mask |= SIG_MASK_REQUIRE_HTTP_STATE;
```
Comment (inliniac) [inline src/detect.c]: I think this is wrong. It sets a condition (rule requires http AND smtp state) that can't be true for a packet/flow. In such cases you can just not set the condition at all. (+ add comment)

File: suricata.yaml [hunk 7]
```diff
@@ -0,0 +1,1483 @@
+%YAML 1.1
```
Comment (inliniac) [inline suricata.yaml]: This shouldn't be in the patch.

Comment (inliniac) [inline suricata.yaml]: Updates to the yaml go into yaml.in

**Comment by inliniac** [comment]
Review complete, please address the issues and submit a new PR.

File: src/app-layer-smtp.c (hunk 1 above)
Comment (glongo) [inline src/app-layer-smtp.c]: ack

File: src/app-layer-smtp.c (hunk 2 above)
Comment (glongo) [inline src/app-layer-smtp.c]: ack

File: src/detect-file-data.c (hunk 3 above)
Comment (glongo) [inline src/detect-file-data.c]: ack

File: src/detect-file-data.c (hunk 4 above)
Comment (glongo) [inline src/detect-file-data.c]: ack

File: src/detect-file-data.c (hunk 5 above)
Comment (glongo) [inline src/detect-file-data.c]: ack

File: suricata.yaml (hunk 7 above)
Comment (glongo) [inline suricata.yaml]: ack

**Pull request closed without merging**